crates/e2e/src/backend.rs

```diff
@@ -92,67 +103,173 @@ pub trait ContractsBackend<E: Environment> {
+    }
+}
+
+#[async_trait]
+pub trait BuilderClient<E: Environment>: ContractsBackend<E> {
+    /// Submits an instantiate call with a gas margin.
+    async fn instantiate_with_gas_margin<
+        Contract: Clone,
+        Args: Send + Sync + Encode + Clone,
+        R,
+    >(
         &mut self,
         contract_name: &str,
         caller: &Keypair,
-        constructor: CreateBuilderPartial<E, Contract, Args, R>,
+        constructor: &mut CreateBuilderPartial<E, Contract, Args, R>,
         value: E::Balance,
+        margin: Option<u64>,
         storage_deposit_limit: Option<E::Balance>,
     ) -> Result<InstantiationResult<E, Self::EventLog>, Self::Error>;
 
-    /// Dry run contract instantiation.
-    async fn instantiate_dry_run<Contract, Args: Send + scale::Encode, R>(
+    /// Submits an instantiate call with a raw gas limit.
+    async fn instantiate_with_gas_limit<
+        Contract: Clone,
+        Args: Send + Sync + Encode + Clone,
+        R,
+    >(
         &mut self,
         contract_name: &str,
         caller: &Keypair,
-        constructor: CreateBuilderPartial<E, Contract, Args, R>,
+        constructor: &mut CreateBuilderPartial<E, Contract, Args, R>,
         value: E::Balance,
+        gas_limit: Weight,
         storage_deposit_limit: Option<E::Balance>,
-    ) -> ContractInstantiateResult<E::AccountId, E::Balance, ()>;
+    ) -> Result<InstantiationResult<E, Self::EventLog>, Self::Error>;
 
-    /// The function subsequently uploads and instantiates an instance of the contract.
-    ///
-    /// This function extracts the Wasm of the contract for the specified contract.
-    ///
-    /// Calling this function multiple times should be idempotent, the contract is
-    /// newly instantiated each time using a unique salt. No existing contract
-    /// instance is reused!
-    async fn upload(
+    /// Submits an call with a gas margin.
+    async fn call_with_gas_margin<Args: Sync + Encode + Clone, RetType: Send + Decode>(
         &mut self,
-        contract_name: &str,
         caller: &Keypair,
+        message: &CallBuilderFinal<E, Args, RetType>,
+        value: E::Balance,
+        margin: Option<u64>,
         storage_deposit_limit: Option<E::Balance>,
-    ) -> Result<UploadResult<E, Self::EventLog>, Self::Error>;
+    ) -> Result<CallResult<E, RetType, Self::EventLog>, Self::Error>
+    where
+        CallBuilderFinal<E, Args, RetType>: Clone;
 
-    /// Executes a `call` for the contract at `account_id`.
+    /// Submits an call with a raw gas limit.
+    async fn call_with_gas_limit<Args: Sync + Encode + Clone, RetType: Send + Decode>(
+        &mut self,
+        caller: &Keypair,
+        message: &CallBuilderFinal<E, Args, RetType>,
+        value: E::Balance,
+        gas_limit: Weight,
+        storage_deposit_limit: Option<E::Balance>,
+    ) -> Result<CallResult<E, RetType, Self::EventLog>, Self::Error>
+    where
+        CallBuilderFinal<E, Args, RetType>: Clone;
```

Comment:
Can't we get rid of all these and just use `bare_call` and `bare_instantiate` from the builders?

Reply:
Nope. Specifically because of https://github.com/paritytech/ink/blob/3d8f9ba4f6681db51a9cce3c599110d78d1c9606/crates/e2e/src/subxt_client.rs#L602

Reply:
My suggestion: https://github.com/paritytech/ink/pull/1938